# One-to-Many and Many-to-Many Joins

In [1]:
import sqlite3
import pandas as pd

conn = sqlite3.connect("data.sqlite")

## One to One Relationships

In [4]:
q = """--sql
SELECT firstName,LastName ,email
FROM employees 
WHERE jobtitle = "Sales Rep";
"""

df = pd.read_sql(q,conn)
print("Number of results :", len(df))

Number of results : 17


In [5]:
df.head()

firstName   lastName                            email
0    Leslie   Jennings   ljennings@classicmodelcars.com
1    Leslie   Thompson   lthompson@classicmodelcars.com
2     Julie   Firrelli   jfirrelli@classicmodelcars.com
3     Steve  Patterson  spatterson@classicmodelcars.com
4  Foon Yue      Tseng      ftseng@classicmodelcars.com

### Cities For Sales rep Employees

In [9]:
q = """
SELECT firstName, lastName ,email,city
FROM employees
JOIN offices
    USING(officeCode)
  WHERE jobtitle = "Sales Rep";
"""
df = pd.read_sql(q,conn)
print("Number of results:", len(df))

Number of results: 17


In [10]:
df.head()

firstName   lastName                            email           city
0    Leslie   Jennings   ljennings@classicmodelcars.com  San Francisco
1    Leslie   Thompson   lthompson@classicmodelcars.com  San Francisco
2     Julie   Firrelli   jfirrelli@classicmodelcars.com         Boston
3     Steve  Patterson  spatterson@classicmodelcars.com         Boston
4  Foon Yue      Tseng      ftseng@classicmodelcars.com            NYC

In the above cases the `JOIN` does not increase the number of records at all.This shows the One To one Relationship in that the records do not increase or decrease

## One To Many Relationship

In [11]:
q = """
SELECT productLine, textDescription
FROM productlines
;
"""
df = pd.read_sql(q, conn)
print("Number of results:", len(df))

Number of results: 7


Joining with Products

In [13]:
q = """ 
SELECT productLine,textDescription,productVendor, productDescription
FROM productlines
    JOIN products
    USING(productLine);
"""
df = pd.read_sql(q, conn)
print("Number of results:", len(df))

Number of results: 110


In [14]:
df.head()

productLine                                    textDescription  \
0  Classic Cars  Attention car enthusiasts: Make your wildest c...   
1  Classic Cars  Attention car enthusiasts: Make your wildest c...   
2  Classic Cars  Attention car enthusiasts: Make your wildest c...   
3  Classic Cars  Attention car enthusiasts: Make your wildest c...   
4  Classic Cars  Attention car enthusiasts: Make your wildest c...   

              productVendor                                 productDescription  
0     Autoart Studio Design  Hood, doors and trunk all open to reveal highl...  
1  Carousel DieCast Legends  Features include opening and closing doors. Co...  
2  Carousel DieCast Legends  The operating parts of this 1958 Chevy Corvett...  
3  Carousel DieCast Legends  This diecast model of the 1966 Shelby Cobra 42...  
4   Classic Metal Creations  1957 die cast Corvette Convertible in Roman Re...

In the above code we can see that the join is used to increase the 7 enties to 110 entries so showing the one to many relationship between the tables

## Many to Many Relationships

In [15]:
q = """ 
SELECT * 
FROM offices;
"""
df = pd.read_sql(q, conn)
print('Number of results:', len(df))

Number of results: 7


In [16]:
q = """
SELECT *
FROM customers
;
"""

df = pd.read_sql(q, conn)
print('Number of results:', len(df))

Number of results: 122


joining the Offices and The Customers

In [17]:
q= """ 
SELECT *
FROM offices
JOIN customers
    USING(state)
    ;"""
df = pd.read_sql(q, conn)
print('Number of results:', len(df))

Number of results: 245


In [18]:
df.head()

officeCode           city            phone       addressLine1 addressLine2  \
0          1  San Francisco  +1 650 219 4782  100 Market Street    Suite 300   
1          1  San Francisco  +1 650 219 4782  100 Market Street    Suite 300   
2          1  San Francisco  +1 650 219 4782  100 Market Street    Suite 300   
3          1  San Francisco  +1 650 219 4782  100 Market Street    Suite 300   
4          1  San Francisco  +1 650 219 4782  100 Market Street    Suite 300   

  state country postalCode territory customerNumber  ... contactLastName  \
0    CA     USA      94080        NA            124  ...          Nelson   
1    CA     USA      94080        NA            129  ...          Murphy   
2    CA     USA      94080        NA            161  ...       Hashimoto   
3    CA     USA      94080        NA            205  ...           Young   
4    CA     USA      94080        NA            219  ...           Young   

  contactFirstName       phone               addressLine1 addressLine2  \
0            Susan  4155551450            5677 Strong St.                
1            Julie  6505555787  5557 North Pendale Street                
2             Juri  6505556809          9408 Furth Circle                
3            Julie  6265557265         78934 Hillside Dr.                
4             Mary  3105552373           4097 Douglas Av.                

            city postalCode country salesRepEmployeeNumber creditLimit  
0     San Rafael      97562     USA                   1165   210500.00  
1  San Francisco      94217     USA                   1165    64600.00  
2     Burlingame      94217     USA                   1165    84600.00  
3       Pasadena      90003     USA                   1166    90700.00  
4       Glendale      92561     USA                   1166    11000.00  

[5 rows x 21 columns]

Whenever you write a SQL query, make sure you understand the unit of analysis you are trying to use. Getting more data from the database is not always better! The above query might make sense as a starting point for something like "what is the ratio of customers to offices in each state", but it's not there yet. Many-to-many joins can be useful, but it's important to be strategic and understand what you're really asking for.